obj/objects.cpp

```diff
@@ -1,315 +1,266 @@
-        area_ = vecA.crossProduct(vecB).magnitude() / 2;
-    }
-
-    void TriangleObj::recalculateNormal()
-    {
-        core::Vec3 vecA = point1_ - point2_;
-        core::Vec3 vecB = point1_ - point3_;
-        core::Vec3 perpendicular = vecA.crossProduct(vecB);
-        normal_ = perpendicular.normalize();
-    }
-
-    bool TriangleObj::arePointsValid()
-    {
-        if (point1_ == point2_ || point1_ == point3_ || point2_ == point3_)
-        {
-            std::stringstream ss;
-            ss << "TriangleObj arguments error. You cannot have duplications of the same point in constructor 1: " << point1_ << ", 2: " << point2_ << ", 3: " << point3_;
-            throw std::invalid_argument(ss.str());
-        }
-
-        core::Vec3 alpha = point1_ - point2_;
-        core::Vec3 beta = point1_ - point3_;
-
-        if (alpha.crossProduct(beta) == core::Vec3(0, 0, 0))
-        {
-            throw std::invalid_argument("Triangle object couldn't be constructed. You cannot have all points of the object linedup straight");
-        }
-        return true;
-    }
-
-    // GETTERS AND SETTERS
-    core::Vec3 TriangleObj::point1() const
-    {
-        return point1_;
-    }
-    void TriangleObj::setPoint1(const core::Vec3 &point)
-    {
-        this->point1_ = point;
-    }
-
-    core::Vec3 TriangleObj::point2() const
-    {
-        return point2_;
-    }
-    void TriangleObj::setPoint2(const core::Vec3 &point)
-    {
-        this->point2_ = point;
-    }
-
-    core::Vec3 TriangleObj::point3() const
-    {
-        return point3_;
-    }
-    void TriangleObj::setPoint3(const core::Vec3 &point)
-    {
-        this->point3_ = point;
-    }
+namespace objects {
+
+void Object::setOrigin(const core::Vec3 &origin) { origin_ = origin; }
+core::Vec3 Object::getOrigin() const { return origin_; }
+
+core::Vec3 Sphere::normal(const core::Vec3 &surfacePoint) const {
+  return (surfacePoint - getOrigin()).normalize();
+}
+
+Sphere::Sphere(const core::Vec3 &origin, float rad) {
+  this->setOrigin(origin);
+  radius_ = rad;
+}
+
+bool Sphere::hitObject(const core::Ray &ray, float freq,
+                       core::RayHitData *hitData) {
+  core::Vec3 rVec3 = ray.getOrigin() - this->getOrigin();
+
+  // this calculates variables that are neccesary to calculate times at which
+  // ray hits the object
+  float beta = 2 * rVec3.scalarProduct(ray.getDirection());
+  float gamma = rVec3.scalarProduct(rVec3) - radius_ * radius_;
+  float discriminant = beta * beta - 4 * gamma;
+
+  // when hit of the sphere never occur:
+  if (discriminant < 0) {
+    return false;
+  }
+
+  float temp = std::sqrt(discriminant);
+  // times that ray hits sphere at:
+  float time1 = (-beta - temp) / 2;
+  float time2 = (-beta + temp) / 2;
+
+  // Ray inside sphere. kAccurracy is for the case that ray is on the edge
+  // of the sphere. time < 0 - hit behind ray, time > 0 - hit front of ray
+  if (time1 < -constants::kAccuracy && time2 > constants::kAccuracy) {
```

Comment:
Im not sure if timeLow is a hit time clsoer to the Origin point. If (-beta) is less then zero it can be actually futher away from origin ... but the code looks very clean :D

Reply:
I made my comments about it. I really like idea of using timeLow > 0 ? timeLow : timeHigh - very space efficient stuff :D

Reply:
Ah right, about the -beta < 0, but those cases are discarded because they fall into the first conditions right?

Reply:
eventually yes, but this implementation works no matter what beta value is, so I felt like this way of explanation that I put is a little bit more precise ;)

Reply:
Not sure if I get it:
1. my impl. works mo matter what beta value is
2. I did not see any changes you made (that is weird)
3. if you actually did not do any changes, then I disagree that explanations/comments you have in the code are more precise, or readable, because you do not explain any details of what time1 and time2 is, what is their relation to each other, and naming is generic.

Reply:
this is weird then, because I copied your code and changed only comments ...